### PR TITLE
When a car trip violates a congestion cap, reroute around the problem

### PR DESCRIPTION
**Pull request by dabreegster**
instead of aborting the trip.

The demo: With a cap on these roads:
edits

Throughput layer now shows higher traffic on the detour, instead of a bunch of aborted trips:
rerouted

The edits:
[crazy_cap.json.txt](https://github.com/dabreegster/abstreet/files/5205721/crazy_cap.json.txt)

@michaelkirk

### CODE REVIEW

File: sim/src/cap.rs [hunk 1]
```diff
@@ -1,16 +1,9 @@
 use crate::{CarID, VehicleType};
 use geom::{Duration, Time};
-use map_model::{LaneID, Map, Path, PathConstraints, PathStep};
+use map_model::{LaneID, Map, Path, PathConstraints, PathRequest, PathStep};
 use serde::{Deserialize, Serialize};
 use std::collections::{BTreeMap, BTreeSet};
 
-// TODO When do we increase the counter for a zone, and when do we enforce the check? If we check
```
Comment (dabreegster) [inline sim/src/cap.rs:7]: I took out the old system, which bumped the counter when a car actually entered the zone. It sounds like GLT wants to have some cars reserve way in advance (for the use case of regular commuters) and some around when they start their trip.

File: sim/src/cap.rs [hunk 2]
```diff
@@ -99,6 +74,33 @@ impl CapSimState {
...
+        car: CarID,
+        map: &Map,
+    ) -> Option<Path> {
+        if self.allow_trip(now, car, &path) {
+            return Some(path);
+        }
+
+        // TODO Make the responses configurable: abort the trip, reroute, delay an hour, switch
+        // modes. Where should this policy be specified? Is it simulation-wide?
+
```
Comment (dabreegster) [inline sim/src/cap.rs:91]: Building `avoid_lanes` and/or the dijkstra is SLOW. Extremely noticeable in debug mode. Will think about optimizations after it's a bit more clear what direction this project takes. Some optimizations could include rebuilding the contraction hierarchies every time a zone is filled/becomes available again. That seems intense, but it should be num_zones * 24 at most.

Comment (michaelkirk) [reply sim/src/cap.rs:93]: I could imagine a "live edit" scenario which might be further reason to make this run fast one day.

**Review by michaelkirk — APPROVED** [review]
LGTM

I ran it and things seem to work for me too:

Screen Shot 2020-09-11 at 11 18 25 AM

**Pull request merged**